File: juno/strategies/mamacx.py
import operator
from decimal import Decimal

from juno import Advice, Candle, indicators
from juno.constraints import Int, Pair, Uniform
from juno.indicators import MA, Ema
from juno.utils import get_module_type

from .strategy import Meta, MidTrendPolicy, Strategy, ma_choices


# Moving average moving average crossover.
class MAMACX(Strategy):
    @staticmethod
    def meta() -> Meta:
        return Meta(
            constraints={
                ('short_period', 'long_period'): Pair(Int(1, 100), operator.lt, Int(2, 101)),
                'neg_threshold': Uniform(Decimal('-1.000'), Decimal('-0.100')),
                'pos_threshold': Uniform(Decimal('+0.100'), Decimal('+1.000')),
                'persistence': Int(0, 10),
                'short_ma': ma_choices,
                'long_ma': ma_choices,
            }
        )

    _short_ma: MA
    _long_ma: MA
    _neg_threshold: Decimal
    _pos_threshold: Decimal

    def __init__(
        self,
        short_period: int,
        long_period: int,
        neg_threshold: Decimal,
        pos_threshold: Decimal,
        persistence: int = 0,
        short_ma: str = Ema.__name__.lower(),
        long_ma: str = Ema.__name__.lower(),
    ) -> None:
        self._short_ma = get_module_type(indicators, short_ma)(short_period)
        self._long_ma = get_module_type(indicators, long_ma)(long_period)

        super().__init__(
            maturity=max(self._long_ma.maturity, self._short_ma.maturity),
            persistence=persistence,
            mid_trend_policy=MidTrendPolicy.IGNORE,
        )
        self.validate(
            short_period, long_period, neg_threshold, pos_threshold, persistence, short_ma, long_ma
        )

        self._short_ma = get_module_type(indicators, short_ma)(short_period)
        self._long_ma = get_module_type(indicators, long_ma)(long_period)
        self._neg_threshold = neg_threshold
        self._pos_threshold = pos_threshold

    def tick(self, candle: Candle) -> Advice:
        self._short_ma.update(candle.close)
        self._long_ma.update(candle.close)

        if self.mature:
            diff = (
                100
                * (self._short_ma.value - self._long_ma.value)
                / ((self._short_ma.value + self._long_ma.value) / 2)
            )

            if diff > self._pos_threshold:
                return Advice.LONG
            elif diff < self._neg_threshold:
                return Advice.SHORT

        return Advice.NONE

File: juno_rs/strategies/mamacx.rs
use std::cmp::{max, min};

use crate::{
    indicators::{ma_from_adler32, MA},
    strategies::{combine, MidTrend, Persistence, Strategy},
    Advice, Candle,
};

pub struct MAMACX {
    short_ma: Box<dyn MA>,
    long_ma: Box<dyn MA>,
    neg_threshold: f64,
    pos_threshold: f64,
    mid_trend: MidTrend,
    persistence: Persistence,
    t: u32,
    t1: u32,
}

impl MAMACX {
    pub fn new(
        short_period: u32,
        long_period: u32,
        neg_threshold: f64,
        pos_threshold: f64,
        persistence: u32,
        short_ma: u32,
        long_ma: u32,
    ) -> Self {
        let short_ma = ma_from_adler32(short_ma, short_period);
        let long_ma = ma_from_adler32(long_ma, long_period);
        let t1 = max(long_ma.maturity(), short_ma.maturity());
        Self {
            short_ma,
            long_ma,
            mid_trend: MidTrend::new(true),
            persistence: Persistence::new(persistence, false),
            neg_threshold,
            pos_threshold,
            t: 0,
            t1,
        }
    }
}

impl Strategy for MAMACX {
    fn update(&mut self, candle: &Candle) -> Advice {
        self.short_ma.update(candle.close);
        self.long_ma.update(candle.close);

        let mut advice = Advice::None;
        if self.t == self.t1 {
            let diff = 100.0 * (self.short_ma.value() - self.long_ma.value())
                / ((self.short_ma.value() + self.long_ma.value()) / 2.0);

            if diff > self.pos_threshold {
                advice = Advice::Long;
            } else if diff < self.neg_threshold {
                advice = Advice::Short;
            }

            advice = combine(
                self.mid_trend.update(advice),
                self.persistence.update(advice),
            );
        }

        self.t = min(self.t + 1, self.t1);
        advice
    }
}
